modules/signals/src/signal-state.ts

```diff
@@ -0,0 +1,34 @@
+import { signal, WritableSignal } from '@angular/core';
+import { toDeepSignal } from './deep-signal';
+import { defaultEqual } from './select-signal';
+import {
+  NotAllowedStateCheck,
+  SignalState,
+  SignalStateUpdate,
+} from './signal-state-models';
+
+export function signalState<State extends Record<string, unknown>>(
+  initialState: State & NotAllowedStateCheck<State>
+): SignalState<State> {
+  const stateSignal = signal(initialState as State, { equal: defaultEqual });
+  const deepSignal = toDeepSignal(stateSignal.asReadonly());
+  (deepSignal as SignalState<State>).$update =
+    signalStateUpdateFactory(stateSignal);
+
+  return deepSignal as SignalState<State>;
+}
+
+export function signalStateUpdateFactory<State extends Record<string, unknown>>(
```

Comment:
Nit:

```suggestion
function signalStateUpdateFactory<State extends Record<string, unknown>>(
```

Reply:
I exported it, because it will be used in the `signalStore` implementation. 😅